### PR TITLE
Fast/warp sync fixes

### PR DESCRIPTION
**Pull request by arkpar**
This fixes a few issues with fast/warp sync, in particular related to restarting interrupted fast sync.

1. When fast sync is interrupted and the node is restarted, babe would attempt to read the configuration from the latest block, instead of the latest available state.
2. When fast sync is restarted, db would re-finalize genesis block, even if better headers exist.
3. After fast sync is complete, client might attempt to re-import existing states when doing reorgs. This is fixed in client.rs. Now only blocks with no state (i.e. headers only) may be re-imported. I've also added a check to `state-db` to better detect such error in the future.
4. When downloading history blocks, the start block was not validated properly with `validate_blocks`, which may lead to download stalling when the other side does not return all requested blocks in the response.

Fixes https://github.com/paritytech/polkadot/issues/4587

### CODE REVIEW

File: client/service/src/client/client.rs [hunk 1]
```diff
@@ -1728,9 +1728,8 @@ where
 			.block_status(&BlockId::Hash(hash))
 			.map_err(|e| ConsensusError::ClientImport(e.to_string()))?
 		{
-			BlockStatus::InChainWithState | BlockStatus::Queued if !import_existing =>
+			BlockStatus::InChainWithState | BlockStatus::Queued =>
```
Comment (bkchr) [inline client/service/src/client/client.rs:1731]: This will break Cumulus.

File: client/db/src/lib.rs [hunk 2]
```diff
@@ -3290,4 +3292,16 @@ pub(crate) mod tests {
...
+	#[test]
+	fn test_import_existing_state_fails() {
+		let backend: Backend<Block> = Backend::new_test(10, 10);
+		insert_block(&backend, 0, Default::default(), None, Default::default(), vec![], None)
+			.unwrap();
+
+		let result =
+			insert_block(&backend, 0, Default::default(), None, Default::default(), vec![], None);
+
+		assert!(result.is_err());
```
Comment (bkchr) [inline client/db/src/lib.rs:3305]: Can you not directly check for the error variant?

File: client/consensus/babe/src/lib.rs [hunk 3]
```diff
@@ -347,7 +347,11 @@ impl Config {
 	{
 		trace!(target: "babe", "Getting slot duration");
 
-		let best_block_id = BlockId::Hash(client.usage_info().chain.best_hash);
+		let mut best_block_id = BlockId::Hash(client.usage_info().chain.best_hash);
+		if client.usage_info().chain.finalized_state.is_none() {
+			debug!(target: "babe", "No finalized state not available. Reading config from genesis");
+			best_block_id = BlockId::Hash(client.usage_info().chain.genesis_hash);
```
Comment (bkchr) [inline client/consensus/babe/src/lib.rs:353]: If we are still fast/warp syncing, why do we set something as best block that isn't yet the block? This could break code in a couple of places. 

The syncing code should set something as best block when it has finished syncing the state etc and not before.

File: client/service/src/client/client.rs (hunk 1 above)
Comment (arkpar) [reply client/service/src/client/client.rs:1731]: Cumulus really needs just to move the best block pointer without reimporting the whole thing. Surprisingly all tests still pass in cumulus with this change.

Comment (bkchr) [reply client/service/src/client/client.rs:1731]: > move the best block pointer without reimporting the whole thing

How?

File: client/consensus/babe/src/lib.rs (hunk 3 above)
Comment (arkpar) [reply client/consensus/babe/src/lib.rs:353]: Not doing so will break code in way more places. First, I should note that only fast sync does that. Warp sync sets the best block only after the full state has been imported. As for fast sync, it downloads the header chains and sets the best header based on the chain selection rule. It needs to verify full header chain before downloading the state, because without warp proofs there's no other way to check if the state is valid. The codebase already supported header-only chain sync, as this is how the light client used to operate before it was removed. The headers as verified and finalised as soon as they are downloaded. In theory we could download full chain and then finalise it at once, but this is too big of a refactoring.

File: client/service/src/client/client.rs (hunk 1 above)
Comment (arkpar) [reply client/service/src/client/client.rs:1731]: Client should expose `set_head` function that does that. It is really out of the scope of this PR so I'll work around it for  now and create a separate issue.

Comment (arkpar) [reply client/service/src/client/client.rs:1731]: Cumulus is fine with this change because it does not even call this function. Furthermore cumulus does not attempt to import the same block state twice. This is what this PR prevents from doing.

File: client/db/src/lib.rs [hunk 4]
```diff
@@ -477,7 +477,6 @@ impl<Block: BlockT> BlockchainDb<Block> {
 		let mut meta = self.meta.write();
 		if number.is_zero() {
 			meta.genesis_hash = hash;
-			meta.finalized_hash = hash;
```
Comment (bkchr) [inline client/db/src/lib.rs:480]: Why is that removed?

File: client/db/src/lib.rs [hunk 5]
```diff
@@ -1347,11 +1346,6 @@ impl<Block: BlockT> Backend<Block> {
 			}
 
 			if number.is_zero() {
-				transaction.set_from_vec(
```
Comment (bkchr) [inline client/db/src/lib.rs:1350]: And this?

File: client/network/src/protocol/sync.rs [hunk 6]
```diff
@@ -1095,15 +1095,17 @@ impl<B: BlockT> ChainSync<B> {
...
 						peer.state = PeerSyncState::Available;
 						if let Some(gap_sync) = &mut self.gap_sync {
 							gap_sync.blocks.clear_peer_download(who);
-							validate_blocks::<B>(&blocks, who, Some(request))?;
-							gap_sync.blocks.insert(start_block, blocks, who.clone());
+							if let Some(start_block) =
+								validate_blocks::<B>(&blocks, who, Some(request))?
+							{
+								gap_sync.blocks.insert(start_block, blocks, who.clone());
+							}
```
Comment (bkchr) [inline client/network/src/protocol/sync.rs:1106]: Could we have a test for this?

File: client/consensus/babe/src/lib.rs (hunk 3 above)
Comment (bkchr) [reply client/consensus/babe/src/lib.rs:353]: Could we then maybe make `best_hash()` a function that is doing this `client.usage_info().chain.finalized_state.is_none()` internally? And also provide some sort of "prospective_best_head"? And `best_hash()` and `prospective_best_head` differ as long as we are doing a fast sync?

File: client/db/src/lib.rs (hunk 4 above)
Comment (arkpar) [reply client/db/src/lib.rs:480]: This is not really needed. Genesis is explicitly finalized in `try_commit_operation` anyway. It is also wrong because we don't want to re-finalize genesis when restarting interrupted fast sync.

File: client/db/src/lib.rs (hunk 5 above)
Comment (arkpar) [reply client/db/src/lib.rs:1350]: This is wrong because we don't want to re-finalize genesis when restarting interrupted fast sync. When doing fast/warp  sync we don't really commit genesis state, because it won't be properly pruned when a recent state is imported. So genesis is not committed so that it does not leave any junk in the database. However, when the fast sync is restarted, we need to populate genesis again, because babe needs to read configuration from it. When this happens we also have some of the headers synced and finalized so we don't want to re-finalize genesis again. Basically the changes here remove unconditional genesis finalization and make sure it is finalized only when being actually committed for the first time.

File: client/consensus/babe/src/lib.rs (hunk 3 above)
Comment (arkpar) [reply client/consensus/babe/src/lib.rs:353]: What would `prospective_best_head` return?
The database contains blocks (headers). Some of them may have state, some may not. State may be not available not just because of the fast sync, but pruning, header-only sync, the light client mode. Although the last was recently removed.
`best_hash` is supposed to return best header hash, not the best available state. Unfortunately the database doe s not keep track of the best available state, only of the best finalized state as it is much more useful. Here I supposed babe could always read configuration from the best avialable finalized state (or genesis if it is unavailable), but I'm not sure of the implications of such a change.

File: client/network/src/protocol/sync.rs (hunk 6 above)
Comment (arkpar) [reply client/network/src/protocol/sync.rs:1106]: Added a test

**Review by bkchr — APPROVED** [review]

File: client/db/src/lib.rs (hunk 5 above)
Comment (bkchr) [reply client/db/src/lib.rs:1350]: Ty for the explanation.

File: client/network/src/protocol/sync.rs (hunk 6 above)
Comment (bkchr) [reply client/network/src/protocol/sync.rs:1106]: Ty

File: client/consensus/babe/src/lib.rs (hunk 3 above)
Comment (bkchr) [reply client/consensus/babe/src/lib.rs]: Okay

**Review by cheme — APPROVED** [review]

File: client/consensus/babe/src/lib.rs [hunk 7]
```diff
@@ -347,7 +347,11 @@ impl Config {
 	{
 		trace!(target: "babe", "Getting slot duration");
 
-		let best_block_id = BlockId::Hash(client.usage_info().chain.best_hash);
+		let mut best_block_id = BlockId::Hash(client.usage_info().chain.best_hash);
+		if client.usage_info().chain.finalized_state.is_none() {
+			debug!(target: "babe", "No finalized state is available. Reading config from genesis");
```
[the remark this answers is not in the document]
Comment (arkpar) [reply client/consensus/babe/src/lib.rs:352]: As I've already mentioned in other comments here, i don't think it would be correct to tie presence of state to fast sync. Substrate as a framework supports header-only sync. Until recently it was used in the light client. Even though light client support is now removed, the header-only sync is still supported and might be useful for some chains. Fast sync is another mode that uses header-only sync as well.

**Comment by arkpar** [comment]
bot merge
